import sys
import os
import sqlite3
from datetime import datetime
from PyQt5.QtWidgets import QApplication, QDialog, QMessageBox
from PyQt5.uic import loadUi
from google import genai
from google.genai import types
from dotenv import load_dotenv

# -----------------
# 1. í™˜ê²½ ë³€ìˆ˜ ë¡œë“œ ë° API í‚¤/DB ì •ë³´ ì„¤ì •
# -----------------
load_dotenv() 
API_KEY = os.environ.get("GEMINI_API_KEY") 

# DB ì„¤ì •: SQLite íŒŒì¼ ê²½ë¡œ
DB_FILE = 'chat_history.db' 

# -----------------
# 2. ë©”ì¸ ì• í”Œë¦¬ì¼€ì´ì…˜ í´ë˜ìŠ¤ (QDialog ìƒì†)
# -----------------
class GeminiChatApp(QDialog):
    def __init__(self):
        super().__init__()
        
        # ğŸš¨ DB ì´ˆê¸°í™”: ì•± ì‹œì‘ ì‹œ DB íŒŒì¼ê³¼ í…Œì´ë¸” ìƒì„±
        self.init_sqlite_db() 
        
        # UI íŒŒì¼ ë¡œë“œ
        try:
            loadUi("gemini.ui", self) 
        except FileNotFoundError:
            QMessageBox.critical(self, "ì˜¤ë¥˜", "Error: 'gemini.ui' íŒŒì¼ì„ ì°¾ì„ ìˆ˜ ì—†ìŠµë‹ˆë‹¤.")
            sys.exit(1)

        # Gemini í´ë¼ì´ì–¸íŠ¸ ë° ì±„íŒ… ì„¸ì…˜ ì´ˆê¸°í™”
        self.chat = None
        self.client = None
        self.model = 'gemini-2.5-flash'
        self.init_gemini_client()
        
        # ğŸš¨ UI ìš”ì†Œ ì—°ê²° ë° ì´ˆê¸° ì„¤ì • (í†µí•© ë¡œì§ ì ìš©)
        
        # 1. ComboBox ì´ˆê¸° ì„¤ì • (ìœ„ì ¯ ì´ë¦„: self.comboBox)
        # UIì— ì¶”ê°€ëœ ë“œë¡­ë‹¤ìš´ ë°•ìŠ¤ì— í•­ëª© ì¶”ê°€
        self.comboBox.addItem("ëŒ€í™”") # ì¸ë±ìŠ¤ 0
        self.comboBox.addItem("ê²€ìƒ‰") # ì¸ë±ìŠ¤ 1
        
        # 2. í•˜ë‚˜ì˜ ë²„íŠ¼ê³¼ Enter í‚¤ë¥¼ í†µí•© í•¸ë“¤ëŸ¬ì— ì—°ê²°
        # ë²„íŠ¼ê³¼ ì—”í„° í‚¤ ëª¨ë‘ handle_actionì„ í˜¸ì¶œí•˜ì—¬ ëª¨ë“œì— ë”°ë¼ ë™ì‘ì„ ë¶„ê¸°í•©ë‹ˆë‹¤.
        self.pushButton.clicked.connect(self.handle_action)
        self.lineEdit.returnPressed.connect(self.handle_action)
        
        # 3. ê¸°ì¡´ì˜ pushButton_2 (ê²€ìƒ‰ ë²„íŠ¼) ì—°ê²°ì€ ì œê±°ë¨

        self.apply_circular_mask()
        
        if self.client:
            self.txtBrowserResult.setText("Gemini AIì—ê²Œ ì§ˆë¬¸ì„ ì…ë ¥í•˜ì„¸ìš”. ë“œë¡­ë‹¤ìš´ìœ¼ë¡œ 'ê²€ìƒ‰' ëª¨ë“œë¥¼ ì„ íƒí•´ ëŒ€í™” ê¸°ë¡ì„ ì°¾ì„ ìˆ˜ ìˆìŠµë‹ˆë‹¤.")

    # -----------------
    # 3. SQLite DB ì´ˆê¸°í™” í•¨ìˆ˜ (í…Œì´ë¸” ìƒì„±)
    # -----------------
    def init_sqlite_db(self):
        try:
            conn = sqlite3.connect(DB_FILE)
            cursor = conn.cursor()
            cursor.execute("""
                CREATE TABLE IF NOT EXISTS chat_history (
                    id INTEGER PRIMARY KEY AUTOINCREMENT,
                    question TEXT NOT NULL,
                    answer TEXT NOT NULL,
                    created_at TEXT NOT NULL
                )
            """)
            conn.commit()
            conn.close()
            print(f"âœ… SQLite DB íŒŒì¼({DB_FILE}) ë° í…Œì´ë¸” ì¤€ë¹„ ì™„ë£Œ.")
        except Exception as e:
            print(f"âŒ SQLite DB ì´ˆê¸°í™” ì‹¤íŒ¨: {e}")

    def init_gemini_client(self):
        if not API_KEY:
            QMessageBox.warning(self, "API ì˜¤ë¥˜", "ğŸš¨ ê²½ê³ : 'GEMINI_API_KEY' í™˜ê²½ ë³€ìˆ˜ê°€ ì„¤ì •ë˜ì§€ ì•Šì•˜ìŠµë‹ˆë‹¤. API ê¸°ëŠ¥ì€ ì‘ë™í•˜ì§€ ì•ŠìŠµë‹ˆë‹¤.")
            self.txtBrowserResult.setText("Gemini API í‚¤ ì˜¤ë¥˜. .env íŒŒì¼ì„ í™•ì¸í•˜ì„¸ìš”.")
            return
            
        try:
            self.client = genai.Client(api_key=API_KEY)
            self.chat = self.client.chats.create(model=self.model)
        except Exception as e:
            error_msg = f"Gemini í´ë¼ì´ì–¸íŠ¸ ì´ˆê¸°í™” ì¤‘ ì˜¤ë¥˜ ë°œìƒ: {e.__class__.__name__}."
            self.txtBrowserResult.setText(error_msg)
            QMessageBox.critical(self, "ì´ˆê¸°í™” ì˜¤ë¥˜", error_msg)
            self.client = None
            self.chat = None

    # -----------------
    # 4. ğŸš¨ í†µí•© ì•¡ì…˜ í•¸ë“¤ëŸ¬ í•¨ìˆ˜ (ComboBox ëª¨ë“œ ë¶„ê¸°)
    # -----------------
    def handle_action(self):
        """ComboBox ì„ íƒì— ë”°ë¼ ëŒ€í™” ë˜ëŠ” ê²€ìƒ‰ í•¨ìˆ˜ë¥¼ í˜¸ì¶œí•˜ëŠ” í†µí•© í•¨ìˆ˜."""
        
        selected_mode = self.comboBox.currentText()
        
        if selected_mode.startswith("ëŒ€í™”"):
            # 'ëŒ€í™”' ëª¨ë“œì¸ ê²½ìš°: Gemini ì§ˆë¬¸ ì „ì†¡
            self.send_question()
            
        elif selected_mode.startswith("ê²€ìƒ‰"):
            # 'ê²€ìƒ‰' ëª¨ë“œì¸ ê²½ìš°: DB ê²€ìƒ‰ ì‹¤í–‰
            self.search_history()
            
        else:
            QMessageBox.warning(self, "ê²½ê³ ", "ëª¨ë“œë¥¼ ì„ íƒí•´ì£¼ì„¸ìš”: ëŒ€í™” ë˜ëŠ” ê²€ìƒ‰")

    # -----------------
    # 5. ì§ˆë¬¸ ì „ì†¡ ë° ë‹µë³€ ìˆ˜ì‹  í•¨ìˆ˜ (handle_actionì—ì„œ í˜¸ì¶œ)
    # -----------------
    def send_question(self):
        if not self.chat: 
            self.txtBrowserResult.append("\n\n[Error]: Gemini ì„œë¹„ìŠ¤ê°€ ì´ˆê¸°í™”ë˜ì§€ ì•Šì•˜ìŠµë‹ˆë‹¤.")
            return

        question = self.lineEdit.text().strip()
        if not question:
            return

        self.lineEdit.clear()
        
        new_entry = f"\n\n[ì§ˆë¬¸]: {question}\n[fox]: ë‹µë³€ì„ ìƒì„±í•˜ëŠ” ì¤‘..."
        self.txtBrowserResult.append(new_entry)
        self.txtBrowserResult.ensureCursorVisible()
        QApplication.processEvents()

        try:
            response = self.chat.send_message(question)
            final_answer = response.text.strip()
            
            # SQLite DB ì €ì¥ í•¨ìˆ˜ í˜¸ì¶œ
            self.save_to_sqlite(question, final_answer)

            current_log = self.txtBrowserResult.toPlainText().rsplit('\n', 1)
            updated_log = current_log[0] + f"\n[fox]: {final_answer}"

            self.txtBrowserResult.setText(updated_log)
            self.txtBrowserResult.ensureCursorVisible()

        except Exception as e:
            error_message = f"API í˜¸ì¶œ ì¤‘ ì˜¤ë¥˜ ë°œìƒ: {type(e).__name__}"
            current_log = self.txtBrowserResult.toPlainText().rsplit('\n', 1)
            updated_log = current_log[0] + f"\n[Error]: {error_message}"

            self.txtBrowserResult.setText(updated_log)
            print(f"API Error: {e}")
            
    # -----------------
    # 6. SQLite ë°ì´í„°ë² ì´ìŠ¤ ì €ì¥ í•¨ìˆ˜
    # -----------------
    def save_to_sqlite(self, question, answer):
        conn = None 
        try:
            current_time = datetime.now().strftime('%Y-%m-%d %H:%M:%S')
            conn = sqlite3.connect(DB_FILE)
            cursor = conn.cursor()
            
            sql = "INSERT INTO chat_history (question, answer, created_at) VALUES (?, ?, ?)"
            cursor.execute(sql, (question, answer, current_time))
            
            conn.commit()
            print(f"âœ… SQLite ì €ì¥ ì„±ê³µ: {current_time}")

        except Exception as e:
            print(f"âŒ SQLite ì €ì¥ ì‹¤íŒ¨: {e}")
        finally:
            if conn:
                conn.close()

    # -----------------
    # 7. SQLite ë°ì´í„°ë² ì´ìŠ¤ ê²€ìƒ‰ í•¨ìˆ˜ (handle_actionì—ì„œ í˜¸ì¶œ)
    # -----------------
    def search_history(self):
        search_term = self.lineEdit.text().strip()
        
        if not search_term:
            self.txtBrowserResult.setText("âš ï¸ ê²€ìƒ‰ì–´ë¥¼ ì…ë ¥í•´ì£¼ì„¸ìš”.")
            return
            
        self.txtBrowserResult.setText(f"ğŸ” '{search_term}' ê²€ìƒ‰ ê²°ê³¼:")
        conn = None
        
        try:
            conn = sqlite3.connect(DB_FILE)
            cursor = conn.cursor()
            
            # SQL LIKE ê²€ìƒ‰ ì¿¼ë¦¬
            search_like = f"%{search_term}%" 
            sql = """
            SELECT created_at, question, answer 
            FROM chat_history 
            WHERE question LIKE ? OR answer LIKE ?
            ORDER BY created_at DESC
            LIMIT 50
            """
            cursor.execute(sql, (search_like, search_like))
            results = cursor.fetchall()

            self.lineEdit.clear() # ê²€ìƒ‰ í›„ ì…ë ¥ì°½ ë¹„ìš°ê¸°

            if results:
                display_text = f"ğŸ” ì´ {len(results)}ê±´ì˜ ê¸°ë¡ì´ ê²€ìƒ‰ë˜ì—ˆìŠµë‹ˆë‹¤ (ê²€ìƒ‰ì–´: '{search_term}'):\n"
                for row in results:
                    # SQLite ê²°ê³¼ëŠ” ì¸ë±ìŠ¤ë¡œ ì ‘ê·¼í•©ë‹ˆë‹¤. (0: created_at, 1: question, 2: answer)
                    date_str = row[0]
                    question_text = row[1]
                    answer_text = row[2]
                    
                    display_text += "\n" + "="*50 + "\n"
                    display_text += f"ë‚ ì§œ: {date_str}\n" 
                    display_text += f"ì§ˆë¬¸: {question_text[:100]}{'...' if len(question_text) > 100 else ''}\n"
                    display_text += f"ë‹µë³€: {answer_text[:200]}{'...' if len(answer_text) > 200 else ''}"
                
                self.txtBrowserResult.setText(display_text)
            else:
                self.txtBrowserResult.setText(f"âŒ '{search_term}'ê³¼ ì¼ì¹˜í•˜ëŠ” ê¸°ë¡ì„ ì°¾ì„ ìˆ˜ ì—†ìŠµë‹ˆë‹¤.")

        except Exception as e:
            self.txtBrowserResult.setText(f"âŒ ê²€ìƒ‰ ì¤‘ ì•Œ ìˆ˜ ì—†ëŠ” ì˜¤ë¥˜ ë°œìƒ: {e}")
        finally:
            if conn:
                conn.close()


    # -----------------
    # 8. QLabel ì›í˜• ë§ˆìŠ¤í¬ ì ìš© í•¨ìˆ˜ 
    # -----------------
    def apply_circular_mask(self):
        profile_label = self.myPic 
        label_size = profile_label.width() 
        
        if label_size == 0:
              label_size = 100 
              
        radius = label_size // 2 
        
        profile_label.setStyleSheet(f"""
            QLabel {{
                border: 3px solid #6699FF;
                border-radius: {radius}px;
                background-color: white; 
                padding: 0px; 
            }}
        """)

# -----------------
# 9. ì• í”Œë¦¬ì¼€ì´ì…˜ ì‹¤í–‰
# -----------------
if __name__ == '__main__':
    app = QApplication(sys.argv)
    window = GeminiChatApp()
    window.show()
    sys.exit(app.exec_())